File: packages/scoring/cbi.py
"""
Copenhagen Burnout Inventory (CBI) — Scoring Module
=====================================================
Reference: Kristensen et al. (2005). The Copenhagen Burnout Inventory: A new tool
for the assessment of burnout. Work & Stress, 19(3), 192-207.

Structure:
  - Personal Burnout  : items 1–6   (frequency scale)
  - Work Burnout      : items 7–13  (degree scale; item 13 reversed)
  - Client Burnout    : items 14–19 (frequency scale)

Response encoding (caller must pass values already mapped to these integers):
  Frequency scale  : Always=100, Often=75, Sometimes=50, Seldom=25, Never/Almost never=0
  Degree scale     : To a very high degree=100, High=75, Somewhat=50, Low=25, Very low=0

Item 13 is REVERSED before averaging Work Burnout.

Score range per subscale: 0–100
Interpretation bands:
  0–49   → Low
  50–74  → Moderate
  ≥75    → High
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Dict, List

# Keys expected in the responses dict
PERSONAL_ITEMS = [f"cbi_{i}" for i in range(1, 7)]    # cbi_1 … cbi_6
WORK_ITEMS     = [f"cbi_{i}" for i in range(7, 14)]   # cbi_7 … cbi_13
CLIENT_ITEMS   = [f"cbi_{i}" for i in range(14, 20)]  # cbi_14 … cbi_19
REVERSED_ITEMS = {"cbi_13"}                            # item 13: high energy = low burnout

ALL_ITEMS = PERSONAL_ITEMS + WORK_ITEMS + CLIENT_ITEMS

VALID_VALUES = {0, 25, 50, 75, 100}

BANDS = [
    (0,  49,  "Low"),
    (50, 74,  "Moderate"),
    (75, 100, "High"),
]


@dataclass
class CBIResult:
    personal_burnout: float
    work_burnout: float
    client_burnout: float
    total: float
    bands: Dict[str, str] = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "subscales": {
                "personal_burnout": {
                    "score": round(self.personal_burnout, 1),
                    "band": self.bands.get("personal_burnout", ""),
                },
                "work_burnout": {
                    "score": round(self.work_burnout, 1),
                    "band": self.bands.get("work_burnout", ""),
                },
                "client_burnout": {
                    "score": round(self.client_burnout, 1),
                    "band": self.bands.get("client_burnout", ""),
                },
            },
            "total": {
                "score": round(self.total, 1),
                "band": self.bands.get("total", ""),
            },
        }


def _get_band(score: float) -> str:
    for low, high, label in BANDS:
        if low <= score <= high:
            return label
    return "High"


def _mean(values: List[float]) -> float:
    return sum(values) / len(values) if values else 0.0


def _reverse(value: int) -> int:
    """Reverse a CBI value: 0↔100, 25↔75, 50↔50."""
    return 100 - value


def validate_responses(responses: Dict[str, int]) -> List[str]:
    """Return a list of validation error messages (empty = valid)."""
    errors: List[str] = []
    missing = [k for k in ALL_ITEMS if k not in responses]
    if missing:
        errors.append(f"Missing items: {missing}")
    for key, val in responses.items():
        if key in {*PERSONAL_ITEMS, *WORK_ITEMS, *CLIENT_ITEMS}:
            if val not in VALID_VALUES:
                errors.append(f"Item {key} has invalid value {val}. Must be one of {VALID_VALUES}.")
    return errors


def score(responses: Dict[str, int]) -> CBIResult:
    """
    Compute CBI subscale scores.

    Parameters
    ----------
    responses : dict
        Mapping of item key → integer value (already on 0/25/50/75/100 scale).
        e.g. {"cbi_1": 75, "cbi_2": 50, …, "cbi_19": 25}

    Returns
    -------
    CBIResult with scores and interpretation bands.
    """
    errors = validate_responses(responses)
    if errors:
        raise ValueError(f"CBI validation failed: {errors}")

    def _value(key: str) -> float:
        val = responses[key]
        return float(_reverse(val) if key in REVERSED_ITEMS else val)

    personal_scores = [_value(k) for k in PERSONAL_ITEMS]
    work_scores     = [_value(k) for k in WORK_ITEMS]
    client_scores   = [_value(k) for k in CLIENT_ITEMS]

    personal = _mean(personal_scores)
    work     = _mean(work_scores)
    client   = _mean(client_scores)
    total    = _mean(personal_scores + work_scores + client_scores)

    result = CBIResult(
        personal_burnout=personal,
        work_burnout=work,
        client_burnout=client,
        total=total,
    )
    result.bands = {
        "personal_burnout": _get_band(personal),
        "work_burnout":     _get_band(work),
        "client_burnout":   _get_band(client),
        "total":            _get_band(total),
    }
    return result

File: packages/scoring/definitions/cbi.ts
/**
 * Copenhagen Burnout Inventory (CBI) — SurveyJS Schema + Scoring Rules
 * ======================================================================
 * 19 items across 3 subscales.
 * Validated instrument: Kristensen et al. (2005), Work & Stress, 19(3).
 *
 * Response values are pre-mapped to CBI integers (0/25/50/75/100) by SurveyJS.
 */

// ─── Shared answer choices ────────────────────────────────────────────────

const FREQUENCY_CHOICES = [
  { value: 100, text: "Always" },
  { value: 75,  text: "Often" },
  { value: 50,  text: "Sometimes" },
  { value: 25,  text: "Seldom" },
  { value: 0,   text: "Never / Almost never" },
];

const DEGREE_CHOICES = [
  { value: 100, text: "To a very high degree" },
  { value: 75,  text: "To a high degree" },
  { value: 50,  text: "Somewhat" },
  { value: 25,  text: "To a low degree" },
  { value: 0,   text: "To a very low degree" },
];

// Item 13 reversed degree choices
const DEGREE_CHOICES_REVERSED = [
  { value: 0,   text: "To a very high degree" },
  { value: 25,  text: "To a high degree" },
  { value: 50,  text: "Somewhat" },
  { value: 75,  text: "To a low degree" },
  { value: 100, text: "To a very low degree" },
];

// ─── SurveyJS schema ──────────────────────────────────────────────────────

export const CBI_SURVEY_SCHEMA = {
  title: "Copenhagen Burnout Inventory",
  description:
    "The following questions ask about how you have been feeling recently. Please choose the answer that best describes your experience.",
  logoPosition: "right",
  showProgressBar: "top",
  showQuestionNumbers: "on",
  questionsOnPageMode: "singlePage",
  pages: [
    {
      name: "personal_burnout",
      title: "Personal Burnout",
      description: "These questions are about how you feel in general.",
      elements: [
        { type: "radiogroup", name: "cbi_1",  title: "How often do you feel tired?",                                choices: FREQUENCY_CHOICES, isRequired: true },
        { type: "radiogroup", name: "cbi_2",  title: "How often are you physically exhausted?",                    choices: FREQUENCY_CHOICES, isRequired: true },
        { type: "radiogroup", name: "cbi_3",  title: "How often are you emotionally exhausted?",                   choices: FREQUENCY_CHOICES, isRequired: true },
        { type: "radiogroup", name: "cbi_4",  title: 'How often do you think "I can\'t take it anymore"?',         choices: FREQUENCY_CHOICES, isRequired: true },
        { type: "radiogroup", name: "cbi_5",  title: "How often do you feel worn out?",                            choices: FREQUENCY_CHOICES, isRequired: true },
        { type: "radiogroup", name: "cbi_6",  title: "How often do you feel weak and susceptible to illness?",     choices: FREQUENCY_CHOICES, isRequired: true },
      ],
    },
    {
      name: "work_burnout",
      title: "Work-Related Burnout",
      description: "These questions are about your work and how you feel about it.",
      elements: [
        { type: "radiogroup", name: "cbi_7",  title: "Is your work emotionally exhausting?",                                         choices: DEGREE_CHOICES,          isRequired: true },
        { type: "radiogroup", name: "cbi_8",  title: "Do you feel burnt out because of your work?",                                   choices: DEGREE_CHOICES,          isRequired: true },
        { type: "radiogroup", name: "cbi_9",  title: "Does your work frustrate you?",                                                 choices: DEGREE_CHOICES,          isRequired: true },
        { type: "radiogroup", name: "cbi_10", title: "Do you feel worn out at the end of the working day?",                           choices: DEGREE_CHOICES,          isRequired: true },
        { type: "radiogroup", name: "cbi_11", title: "Are you exhausted in the morning at the thought of another day at work?",       choices: DEGREE_CHOICES,          isRequired: true },
        { type: "radiogroup", name: "cbi_12", title: "Do you feel that every working hour is tiring for you?",                        choices: DEGREE_CHOICES,          isRequired: true },
        { type: "radiogroup", name: "cbi_13", title: "Do you have enough energy for family and friends during leisure time?",          choices: DEGREE_CHOICES_REVERSED, isRequired: true },
      ],
    },
    {
      name: "client_burnout",
      title: "Client-Related Burnout",
      description: "These questions are about your work with the people you interact with professionally (clients, customers, patients, or similar).",
      elements: [
        { type: "radiogroup", name: "cbi_14", title: "Do you find it hard to work with clients?",                                                  choices: DEGREE_CHOICES,  isRequired: true },
        { type: "radiogroup", name: "cbi_15", title: "Does it drain your energy to work with clients?",                                             choices: DEGREE_CHOICES,  isRequired: true },
        { type: "radiogroup", name: "cbi_16", title: "Do you find it frustrating to work with clients?",                                            choices: DEGREE_CHOICES,  isRequired: true },
        { type: "radiogroup", name: "cbi_17", title: "Do you feel that you give more than you get back when you work with clients?",                 choices: DEGREE_CHOICES,  isRequired: true },
        { type: "radiogroup", name: "cbi_18", title: "Are you tired of working with clients?",                                                      choices: FREQUENCY_CHOICES, isRequired: true },
        { type: "radiogroup", name: "cbi_19", title: "Do you sometimes wonder how long you will be able to continue working with clients?",          choices: FREQUENCY_CHOICES, isRequired: true },
      ],
    },
  ],
};

// ─── Scoring rules (used by API and Python service) ───────────────────────

export const CBI_SCORING_RULES = {
  type: "CBI",
  version: "1.0",
  subscales: {
    personal_burnout: {
      items: ["cbi_1","cbi_2","cbi_3","cbi_4","cbi_5","cbi_6"],
      reversed: [],
      method: "mean",       // mean of pre-mapped values (0/25/50/75/100)
    },
    work_burnout: {
      items: ["cbi_7","cbi_8","cbi_9","cbi_10","cbi_11","cbi_12","cbi_13"],
      reversed: ["cbi_13"], // encoded in choices; already reversed in schema
      method: "mean",
    },
    client_burnout: {
      items: ["cbi_14","cbi_15","cbi_16","cbi_17","cbi_18","cbi_19"],
      reversed: [],
      method: "mean",
    },
  },
  total: {
    method: "mean_of_all",
  },
  bands: [
    { min: 0,  max: 49,  label: "Low",      description: "Burnout is not a significant problem." },
    { min: 50, max: 74,  label: "Moderate",  description: "Some burnout present — monitor and take preventive steps." },
    { min: 75, max: 100, label: "High",      description: "High burnout — intervention is recommended." },
  ],
};
